File: morning_exercises/day15.cpp
#include <memory>
#include <iostream>
#include <optional>

class BSTNode {
    public:
        int val = 0;
        std::shared_ptr<BSTNode> left = nullptr;
        std::shared_ptr<BSTNode> right = nullptr;
    BSTNode(int val): val(val) {}

    static std::optional<int> inOrderSuccessor(std::shared_ptr<BSTNode> root, int k) {
        // if (root==nullptr) return std::nullopt;

        std::optional<int> successor = std::nullopt;

        while (root) {
            if (k<root->val) {
                successor=root->val;
                root=root->left;
            }
            else root=root->right;
        }
        return successor;
    }
};

int main() {
    auto root = std::make_shared<BSTNode>(5);
    root->left = std::make_shared<BSTNode>(3);
    root->right = std::make_shared<BSTNode>(7);
    root->left->left = std::make_shared<BSTNode>(2);
    root->left->right = std::make_shared<BSTNode>(4);

    auto result = BSTNode::inOrderSuccessor(root, 7);
    if (result.has_value()) {
        std::cout << result.value() << "\n";
    } else {
        std::cout << "No successor" << "\n";
    }
    // std::cout << BSTNode::inOrderSuccessor(root, 7) << "\n";
}

// class BSTNode2 {
//     public:
//         int val;
//         std::shared_ptr<BSTNode2> left;
//         std::shared_ptr<BSTNode2> right;

//     BSTNode2(int val = 0): val(val), left(nullptr), right(nullptr) {}
// };

File: morning_exercises/day15.py
from __future__ import annotations

class BSTNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None
    
    @staticmethod
    def inOrderSuccessorLA(root: BSTNode | None, result: dict)->None:
        if (root is None): return

        BSTNode.inOrderSuccessorLA(root.left, result)
        result[root.val]=root.val
        BSTNode.inOrderSuccessorLA(root.right, result)

    @staticmethod
    def successorLA(root: BSTNode | None, k: int)->int:
        result = {}
        BSTNode.inOrderSuccessorLA(root, result)
        keys = result.keys()
        for i in keys:
            if i>k: return result.get(i, 0)

    # @staticmethod
    # def inOrderSuccessor(root: BSTNode | None, result: list, k: int, min: float = float('-inf'), max: float = float('inf'))->None:
        # if ((root is None) or (result[0] is not None)): return

        # if (k<root.val): 
        #     BSTNode.inOrderSuccessor(root.left, result, k, min, root.val)
        #     if result[0] is None:
        #         result[0]=root.val
        
        # if k==root.val:
        #     # result[0]=root.val
        #     return

        # if (k>root.val):
        #     BSTNode.inOrderSuccessor(root.right, result, k, root.val, max)
        #     if result[0] is None:
        #         result[0]=root.val
    
    # @staticmethod
    # def successor(root, k):
    #     result = [None]
    #     BSTNode.inOrderSuccessor(root, result, k)
    #     return result[0]

    @staticmethod
    def inOrderSuccessor(root: BSTNode | None, k: int)->int | None:
        successor = None

        while(root):
            if(root.val>k):
                successor=root.val
                root=root.left
            else: root=root.right
            
        return successor

root = BSTNode(5)
root.left = BSTNode(4)
root.right = BSTNode(7)
root.right.left = BSTNode(6)
root.right.right = BSTNode(8)
root.right.right.right = BSTNode(10)

# print(BSTNode.successorLA(root, 8))
print(BSTNode.inOrderSuccessor(root, 6))
